staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go

```diff
@@ -126,9 +126,19 @@ func BeforeDelete(strategy RESTDeleteStrategy, ctx context.Context, obj runtime.
 			if period >= *objectMeta.GetDeletionGracePeriodSeconds() {
 				return false, true, nil
 			}
+			// Move the existing deletionTimestamp back by existing object.DeletionGracePeriod, then forward by options.DeletionGracePeriod.
+			// This moves the deletionTimestamp back, since the grace period can only be shortened in this code path.
 			newDeletionTimestamp := metav1.NewTime(
 				objectMeta.GetDeletionTimestamp().Add(-time.Second * time.Duration(*objectMeta.GetDeletionGracePeriodSeconds())).
 					Add(time.Second * time.Duration(*options.GracePeriodSeconds)))
+			// Prevent shortening the grace period moving deletionTimestamp into the past
+			if now := metav1Now(); newDeletionTimestamp.Before(&now) {
+				newDeletionTimestamp = now
+				if period != 0 {
+					// if immediate deletion (period==0) was not requested, shorten to the minimum grace period possible
+					period = int64(1)
```

Comment:
working through an example.

1. GracePeriodSeconds set to 20
2. Delete thing thing
3. wait 15s
4. set GracePeriodSeconds to 10
5. Why shouldn't this be treated as immediate deletion since more than 10s has already elapsed?

Reply:
that changes *who* the deleting actor is from whatever observed the resulting `Update` watch event to the API server... I'm trying to avoid a behavior change that dramatic

the intent of the caller here is to shorten the grace period, not to force delete

Reply:
Clarify that intent in the code please.  It wasn't obvious to me and I can easily imagine refactoring it away.

After that, lgtm

Reply:
Updated the comment